File: Prácticas/P2/Ruby/civitas/lib/casilla.rb
# encoding: utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative "tipo_casilla.rb"
require_relative "diario.rb"
require_relative "titulo_propiedad.rb"
require_relative "sorpresa.rb"
require_relative "mazo_sorpresas.rb"

module Civitas
  class Casilla
    attr_reader :nombre, :titulo_propiedad
    
    def initialize()
      init()
      @diario = Diario.instance
    end
    
    def self.ini_descanso(cadena)
      self.new()
      @nombre = cadena
      @tipo = TipoCasilla::DESCANSO
    end
    
    def self.ini_calle(titulo)
      self.new()
      @titulo_propiedad = titulo
      @nombre = titulo.nombre
      @tipo = TipoCasilla::CALLE
      @importe = titulo.precio_compra
    end
    
    def self.ini_impuesto(cantidad, nombre)
      self.new()
      @importe = cantidad
      @nombre = nombre
      @tipo = TipoCasilla::IMPUESTO
    end
    
    def self.ini_juez(num_casilla_carcel,nombre)
      self.new()
      @@carcel = num_casilla_carcel
      @nombre = nombre
      @tipo = TipoCasilla::JUEZ
    end
    
    def self.ini_sorpresa(mazo,nombre)
      self.new()
      @mazo = mazo
      @nombre = nombre
      @tipo = TipoCasilla::SORPRESA
    end
    
    #private_class_method :new
    
    private
    
    def informe(iactual, todos)
      @diario.ocurre_evento("En la casilla " + to_s() + " esta " + todos[iactual].to_s())
    end
    
    def init()
      @nombre = ""
      @importe = 0
    end
    
    
    public
    
    def jugador_correcto(iactual,todos)
      todos.lenght > iactual
    end
    
    def recibe_jugador(iactual,todos)
      
    end
    
    def recibe_jugador_calle(iactual,todos)
      
    end
    
    def recibe_jugador_impuestos(iactual,todos)
      if(jugador_correcto(iactual,todos))
        informe(iactual, todos)
        todos.at(iactual).paga_impuestos(@importe)
      end
    end
    
    def recibe_jugador_juez(iactual,todos)
      if(jugador_correcto(iactual,todos))
        informe(iactual, todos)
        todos.at(iactual).encarcelar(@importe)
      end
    end
    
    def recibe_jugador_encarcelar(iactual,todos)
      
    end
    
    def to_s
      "Nombre #{@nombre}"
    end
    
  end
end

File: Prácticas/P2/Java/civitas/src/main/java/civitas/Casilla.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;

/**
 *
 * @author Jose Padial
 */
public class Casilla {
    private String nombre;
    private float importe;
    private static int carcel;
    private TipoCasilla tipo;
    private TituloPropiedad tituloPropiedad;
    private Sorpresa sorpresa;
    private MazoSorpresas mazo;
    private Diario diario = Diario.getInstance();
    
    Casilla(String cadena){
        init();
        nombre = cadena;
        this.tipo = TipoCasilla.DESCANSO;
    }
    
    Casilla(TituloPropiedad titulo){
        init();
        this.tituloPropiedad = titulo;
        this.nombre = titulo.getNombre();
        this.tipo = TipoCasilla.CALLE;
        this.importe = titulo.getPrecioCompra();
    }
    
    Casilla(float cantidad, String nombre){
        init();
        this.importe = cantidad;
        this.nombre = nombre;
        this.tipo = TipoCasilla.IMPUESTO;
    }
    
    Casilla(int numCasillaCarcel, String nombre){
        init();
        carcel = numCasillaCarcel;
        this.nombre = nombre;
        this.tipo = TipoCasilla.JUEZ;
    }
    
    Casilla(MazoSorpresas mazo, String nombre){
        init();
        this.mazo = mazo;
        this.nombre = nombre;
        this.tipo = TipoCasilla.SORPRESA;
    }

    public String getNombre(){
        return nombre;
    }

    TituloPropiedad getTituloPropiedad() {
        return tituloPropiedad;
    }
    
    private void informe(int iactual, ArrayList<Jugador> todos){
        diario.ocurreEvento("En la casilla " + toString() + " esta " + todos.get(iactual).toString());
    }
    
    private void init(){
        this.nombre = "";
        this.importe = 0;
    }
    
    public boolean juagadorCorrecto(int iactual, ArrayList<Jugador> todos){
        return todos.size() > iactual;
    }
    
    void recibeJugador(int iactual, ArrayList<Jugador> todos){
        
    }
    
    private void recibeJugador_calle(int iactual, ArrayList<Jugador> todos){
        
    }
    
    private void recibeJugador_impuesto(int iactual, ArrayList<Jugador> todos){
        if(juagadorCorrecto(iactual, todos)){
            informe(iactual, todos);
            todos.get(iactual).pagaImpuesto(importe);
        }
    }
    
    private void recibeJugador_juez(int iactual, ArrayList<Jugador> todos){
        if(juagadorCorrecto(iactual, todos)){
            informe(iactual, todos);
            todos.get(iactual).encarcelar(carcel);
        }
    }
    
    private void recibeJugador_sorpresa(int iactual, ArrayList<Jugador> todos){
        
    }

    @Override
    public String toString() {
        return "Casilla{" + "nombre=" + nombre + ", importe=" + importe + ", tipo=" + tipo + ", tituloPropiedad=" + tituloPropiedad + ", sorpresa=" + sorpresa + ", mazo=" + mazo + '}';
    }
    
}
